fabric-api-base/src/testmodClient/java/net/fabricmc/fabric/test/base/client/mixin/MinecraftClientMixin.java

```diff
@@ -0,0 +1,149 @@
+	private void postRunTasks(CallbackInfo ci) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST) {
+			ThreadingImpl.clientCanAcceptTasks = true;
+			ThreadingImpl.enterPhase(ThreadingImpl.PHASE_TEST);
+
+			if (ThreadingImpl.testThread != null) {
+				while (true) {
+					try {
+						ThreadingImpl.CLIENT_SEMAPHORE.acquire();
+					} catch (InterruptedException e) {
+						throw new RuntimeException(e);
+					}
+
+					if (ThreadingImpl.taskToRun != null) {
+						ThreadingImpl.taskToRun.run();
+					} else {
+						break;
+					}
+				}
+			}
+
+			ThreadingImpl.enterPhase(ThreadingImpl.PHASE_TICK);
+
+			Runnable deferredTask = this.deferredTask;
+			this.deferredTask = null;
+
+			if (deferredTask != null) {
+				deferredTask.run();
+			}
+		}
+	}
+
+	@Inject(method = "startIntegratedServer", at = @At("HEAD"), cancellable = true)
+	private void deferStartIntegratedServer(LevelStorage.Session session, ResourcePackManager dataPackManager, SaveLoader saveLoader, boolean newWorld, CallbackInfo ci) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST && ThreadingImpl.taskToRun != null) {
+			// don't start the integrated server (which busywaits) inside a task
+			deferredTask = () -> MinecraftClient.getInstance().startIntegratedServer(session, dataPackManager, saveLoader, newWorld);
+			ci.cancel();
+		}
+	}
+
+	@Inject(method = "startIntegratedServer", at = @At(value = "INVOKE", target = "Ljava/lang/Thread;sleep(J)V", remap = false))
+	private void onStartIntegratedServerBusyWait(CallbackInfo ci) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST) {
+			// give the server a chance to tick too
+			preRunTasks(ci);
+			postRunTasks(ci);
+		}
+	}
+
+	@Inject(method = "disconnect(Lnet/minecraft/client/gui/screen/Screen;Z)V", at = @At("HEAD"), cancellable = true)
+	private void deferDisconnect(Screen disconnectionScreen, boolean transferring, CallbackInfo ci) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST && MinecraftClient.getInstance().getServer() != null && ThreadingImpl.taskToRun != null) {
+			// don't disconnect (which busywaits) inside a task
+			deferredTask = () -> MinecraftClient.getInstance().disconnect(disconnectionScreen, transferring);
+			ci.cancel();
+		}
+	}
+
+	@Inject(method = "disconnect(Lnet/minecraft/client/gui/screen/Screen;Z)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/MinecraftClient;render(Z)V", shift = At.Shift.AFTER))
+	private void onDisconnectBusyWait(CallbackInfo ci) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST) {
+			// give the server a chance to tick too
+			preRunTasks(ci);
+			postRunTasks(ci);
+		}
+	}
+
+	@Inject(method = "getInstance", at = @At("HEAD"))
+	private static void checkThreadOnGetInstance(CallbackInfoReturnable<MinecraftClient> cir) {
+		if (FabricApiAutoTestClient.IS_AUTO_TEST) {
+			Preconditions.checkState(Thread.currentThread() != ThreadingImpl.testThread, "MinecraftClient.getInstance() cannot be called from the test thread");
+		}
+	}
```

Comment:
This is a really nice addition that I'm sure will catch a lot of newbie mistakes. Maybe we should go one step further and also point them in the right direction. Something like:

> MinecraftClient.getInstance() cannot be called from the test thread. You probably want to use runOnClient() or computeOnClient() instead.

Reply:
I didn't want to add this yet because we don't know what the user-facing methods are going to be called

Reply:
Yes, the method names might change later and would then need to be updated in this message, but I don't think the general concept of "lambda that runs stuff on the client thread" and "other lambda that does that and also returns a value" is going away any time soon.

The risk of forgetting to add this later seems worse. Remember, first-time modders / test writers are not used to the way this system works and may not even know that these methods exist. It's not obvious from the current code that the message is unfinished. Consider adding a `// TODO:` comment if you don't want to add the full message yet.